### PR TITLE
Aggregated Signed Signatures on Consensus

### PR DESCRIPTION
**Pull request by davidrusu**
On termination, elders can now create an aggregated signature for each proposal

### CODE REVIEW

File: src/consensus.rs [hunk 1]
```diff
@@ -199,20 +216,48 @@ impl<T: Proposition> Consensus<T> {
...
+                .filter(|v| v.vote.is_super_majority_ballot())
+                .cloned(),
+        );
+        if let Some((props, count_of_sm)) = self
+            .count_votes(&sm_votes)
             .into_iter()
-            .map(|(_, count)| count)
-            .max()
-            .unwrap_or(0);
+            .max_by_key(|(_, c)| *c)
```
Comment (grumbach) [inline src/consensus.rs:232]: Just wondering, is it possible to have multiple max here? What happens in that case? (eg: 3 bad nodes that vote twice in a split network, 5 vote for A, 5 for B)

File: src/consensus.rs [hunk 2]
```diff
@@ -221,7 +266,6 @@ impl<T: Proposition> Consensus<T> {
         self.verify_sig_share(&signed_vote.vote, signed_vote.voter, &signed_vote.sig)?;
         self.validate_vote(&signed_vote.vote)?;
         self.validate_vote_supersedes_existing_vote(signed_vote)?;
-        self.validate_voters_have_not_changed_proposals(signed_vote)?;
```
Comment (grumbach) [inline src/consensus.rs:224]: Why? Is that useless because of the supersede check above?

File: src/vote.rs [hunk 3]
```diff
@@ -69,7 +80,17 @@ impl<T: Proposition> Debug for Vote<T> {
...
     pub fn is_super_majority_ballot(&self) -> bool {
-        matches!(self.ballot, Ballot::SuperMajority(_))
+        matches!(self.ballot, Ballot::SuperMajority { .. })
+    }
+
+    pub fn proposals(&self) -> BTreeSet<T> {
+        match &self.ballot {
+            Ballot::Propose(proposal) => BTreeSet::from_iter([proposal.clone()]),
+            Ballot::Merge(votes) | Ballot::SuperMajority { votes, .. } => {
+                // TAI: use proposals instead of recursing on SuperMajority?
```
Comment (grumbach) [inline src/vote.rs:90]: Sounds like a good optimisation

**Review by grumbach — APPROVED** [review]
LGTM!

File: src/consensus.rs (hunk 1 above)
Comment (davidrusu) [reply src/consensus.rs:232]: You could have multiple max's when we are building up to super-majority. i.e. 1 vote for A, 1 vote for B.

But once we pass the super-majority threshold, only one max will be possible (as long as our assumption of < 1/3 faulty nodes holds.)

File: src/vote.rs (hunk 3 above)
Comment (davidrusu) [reply src/vote.rs:90]: future work!

**Pull request merged**